doc/dev/transforms.rst

```diff
@@ -0,0 +1,594 @@
+
+        if (!isa<QubitUnitaryOp>(parent))
+            return failure();
+
+        QubitUnitaryOp parentOp = cast<QubitUnitaryOp>(parent);
+        ValueRange parentQbs = parentOp.getOutQubits();
+
+        if (qbs.size() != parentQbs.size())
+            return failure();
+
+        for (auto [qb1, qb2] : llvm::zip(qbs, parentQbs))
+            if (qb1 != qb2)
+                return failure();
+
+        return success();
+    }
+
+Note that we have used a couple of functions (like ``getInQubits`` and ``getOutQubits``) from the
+definition of the ``QubitUnitaryOp`` class. Since we define our operations in the declarative ODS
+(tablegen) format, the corresponding C++ classes are automatically generated. This is the definition
+for the ``QubitUnitaryOp`` from the `QuantumOps.td <https://github.com/PennyLaneAI/catalyst/blob/201b0ec6cbec18b6411a876a3c72ba878123e2a1/mlir/include/Quantum/IR/QuantumOps.td#L267>`_
+file:
+
+.. code-block::
+
+    def QubitUnitaryOp : Gate_Op<"unitary"> {
+        let summary = "Apply an arbitrary fixed unitary matrix";
+        let description = [{
+            The `quantum.unitary` operation applies an arbitrary fixed unitary matrix to the
+            state-vector. The arguments are a set of qubits and a 2-dim matrix of complex numbers
+            that represents a Unitary matrix of size 2^(number of qubits) * 2^(number of qubits).
+        }];
+
+        let arguments = (ins
+            2DTensorOf<[Complex<F64>]>:$matrix,
+            Variadic<QubitType>:$in_qubits
+        );
+
+        let results = (outs
+            Variadic<QubitType>:$out_qubits
+        );
+
+        let assemblyFormat = [{
+            `(` $matrix `:` type($matrix) `)` $in_qubits attr-dict `:` type($out_qubits)
+        }];
+    }
+
+MLIR will automatically generate canonical ``get*`` methods for attributes like ``in_qubits``,
+``out_qubits``, and ``matrix``. When in doubt it's best to have a look at the generated C++ files in
+the build folder, named ``QuantumOps.h.inc`` and ``QuantumOps.cpp.inc`` in this instance.
+
+Alright, now that we have the matching part, let's implement the actual transformation via the
+``rewrite`` method. All we need to do is replace the original pattern with the following:
+
+.. code-block::
+
+    QubitUnitary(A, QubitUnitary(B, Q))  -->  QubitUnitary(AxB, Q)
+
+In C++ it will look as follows:
+
+.. code-block:: cpp
+
+    void QubitUnitaryFusion::rewrite(QubitUnitaryOp op, PatternRewriter &rewriter)
+    {
+        ValueRange qbs = op.getInQubits();
+        QubitUnitaryOp parentOp = cast<QubitUnitaryOp>(qbs[0].getDefiningOp());
+
+        Value m1 = op.getMatrix();
+        Value m2 = parentOp.getMatrix();
+
+        linalg::MatmulOp matmul = rewriter.create<linalg::MatmulOp>(op.getLoc(), {m1, m2}, {});
+        Value res = matmul.getResult(0);
+
+        rewriter.updateRootInPlace(op, [&] {
+            op->setOperand(0, res);
+        });
+        rewriter.replaceOp(parentOp, parentOp.getResults());
```

Comment:
It might be worth to explain the `.getResults()`

Reply:
I'm hesitant to explain details of MLIR functions and objects here, as there is too much to go over for the guide to be effective.
What exactly would like to see about `getResults`?